### PR TITLE
Multi environment support

### PR DESCRIPTION
**Pull request by anweiss**
Adds support for declaring a custom service management endpoint. Referencing #40

### CODE REVIEW

File: azure/client.go [hunk 1]
```diff
@@ -36,8 +37,14 @@ func NewAnonymousClient() Client {
 
 //NewClientFromPublishSettingsFile creates a new azure.Client and imports the publish
 //settings from the specified file path.
-func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, error) {
-	client := Client{}
+func NewClientFromPublishSettingsFile(publishSettingsFilePath string, serviceManagementURL ...string) (Client, error) {
```
Comment (ahmetalpbalkan) [inline azure/client.go]: why `...string`?

File: azure/client.go [hunk 2]
```diff
@@ -36,8 +37,14 @@ func NewAnonymousClient() Client {
...
+	var client Client
+	if len(serviceManagementURL) > 0 {
```
Comment (ahmetalpbalkan) [inline azure/client.go]: umm why is this an array?

File: azure/client.go [hunk 3]
```diff
@@ -47,8 +54,13 @@ func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, e
 
 //NewClientFromPublishSettingsFile creates a new azure.Client and imports the publish
 //settings from the specified file path.
-func NewClientFromPublishSettings(subscriptionId string, managementCertificatePath string) (Client, error) {
-	client := Client{}
+func NewClientFromPublishSettings(subscriptionId string, managementCertificatePath string, serviceManagementURL ...string) (Client, error) {
+	var client Client
+	if len(serviceManagementURL) > 0 {
```
Comment (ahmetalpbalkan) [inline azure/client.go]: you just duplicated logic in 2 places. when you do that it means probably you should have it only in one place and these methods should call it.

File: azure/client.go (hunk 2 above)
Comment (anweiss) [inline azure/client.go]: Wasn't sure if was better to use a variadic parameter or create new functions altogether

File: azure/client.go [hunk 4]
```diff
@@ -36,8 +37,14 @@ func NewAnonymousClient() Client {
...
 //NewClientFromPublishSettingsFile creates a new azure.Client and imports the publish
 //settings from the specified file path.
-func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, error) {
-	client := Client{}
+func NewClientFromPublishSettingsFile(publishSettingsFilePath string, serviceManagementURL ...string) (Client, error) {
+	var client Client
+	if len(serviceManagementURL) > 0 {
+		client = Client{serviceManagementURL: serviceManagementURL[0]}
+	} else {
+		client = Client{}
```
Comment (ahmetalpbalkan) [inline azure/client.go]: instead do:
```
var client Client
if url passed
    client.base=passedUrl
else
    client.base=defaultUrl

**Comment by ahmetalpbalkan** [comment]
Please squash changes to 1 commit. I see some intermediate confusing commits that contributes no delta to the PR.

File: azure/client.go (hunk 2 above)
Comment (ahmetalpbalkan) [inline azure/client.go]: Can you give an usage example please? Why should anyone pass multiple URLs to this function?

Comment (anweiss) [inline azure/client.go]: So my use of a variadic parameter was to create an "optional" parameter. You'd never pass multiple values to this. Wasn't sure if it was best practice to do this or create new functions altogether

Comment (ahmetalpbalkan) [inline azure/client.go]: @anweiss please refer to `storage` package to see how I did it :)

Comment (anweiss) [inline azure/client.go]: lol fair enough..will fix this up and squash the commits

**Comment by anweiss** [comment]
squashed commits, cleaned up base URL declaration. Not sure if this is right approach but learning process for me :)

File: azure/virtualmachine/client.go [hunk 5]
```diff
@@ -100,9 +100,7 @@ func (self VirtualMachineClient) CreateAzureVM(azureVMConfiguration *Role, dnsNa
 		return err
 	}
 
-	self.client.WaitAsyncOperation(requestId)
-
-	return nil
+	return self.client.WaitAsyncOperation(requestId)
```
Comment (ahmetalpbalkan) [inline azure/virtualmachine/client.go]: These are fixed already. Please rebase.

File: azure/client.go [hunk 6]
```diff
@@ -45,10 +46,42 @@ func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, e
 	return client, nil
 }
 
-//NewClientFromPublishSettingsFile creates a new azure.Client and imports the publish
-//settings from the specified file path.
-func NewClientFromPublishSettings(subscriptionId string, managementCertificatePath string) (Client, error) {
+// NewClientFromPublishSettingsFile creates a new azure.Client, imports the publish
+// settings from the specified file path, and uses a specific API endpoint.
+func NewClientFromPublishSettingsFile(publishSettingsFilePath string, baseURL string) (Client, error) {
```
Comment (ahmetalpbalkan) [inline azure/client.go]: reuse NewBasicClientFromPublishSettingsFile

File: azure/client.go [hunk 7]
```diff
@@ -45,10 +46,42 @@ func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, e
...
+	err := client.importPublishSettings(subscriptionId, managementCertificatePath)
+	if err != nil {
+		return client, err
+	}
+	return client, nil
+}
+
+// NewClientFromPublishSettings creates a new azure.Client using the specified
+// subscription ID, certificate path, and specific API endpoint
+func NewClientFromPublishSettings(subscriptionId string, managementCertificatePath string, baseURL string) (Client, error) {
```
Comment (ahmetalpbalkan) [inline azure/client.go]: reuse NewBasicClientFromPublishSettings. too much duplication here...

File: azure/client.go [hunk 8]
```diff
@@ -45,10 +46,42 @@ func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, e
...
+	client := Client{}
+	if baseURL == "" {
+		return client, fmt.Errorf("azure: base URL required")
```
Comment (ahmetalpbalkan) [inline azure/client.go]: use errors.New

File: azure/client.go [hunk 9]
```diff
@@ -45,10 +46,42 @@ func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, e
...
 
-//NewClientFromPublishSettingsFile creates a new azure.Client and imports the publish
-//settings from the specified file path.
-func NewClientFromPublishSettings(subscriptionId string, managementCertificatePath string) (Client, error) {
+// NewClientFromPublishSettingsFile creates a new azure.Client, imports the publish
+// settings from the specified file path, and uses a specific API endpoint.
+func NewClientFromPublishSettingsFile(publishSettingsFilePath string, baseURL string) (Client, error) {
 	client := Client{}
+	if baseURL == "" {
+		return client, fmt.Errorf("azure: base URL required")
```
Comment (ahmetalpbalkan) [inline azure/client.go]: errors.New. you don't have a %s or anything really.

File: azure/http.go [hunk 10]
```diff
@@ -159,6 +159,11 @@ func (client *Client) createAzureRequest(url string, requestType string, content
 	var err error
 
 	url = fmt.Sprintf("%s/%s/%s", azureManagementDnsName, client.publishSettings.SubscriptionID, url)
+
+	if client.baseURL != "" {
+		url = fmt.Sprintf("%s/%s/%s", client.baseURL, client.publishSettings.SubscriptionID, url)
```
Comment (ahmetalpbalkan) [inline azure/http.go]: yeah you are adding URL crafting logic to here. this is wrong. all clients must have a `baseURL` – either user-passed or `azureManagementDnsName` and this logic should not be here but at an upper level.

This method should just do c.BaseUrl+"/"+subs+"/+url

**Comment by anweiss** [comment]
not sure how else to consolidate the number of functions here

**Comment by anweiss** [comment]
not sure how else to consolidate the number of functions here

File: azure/client.go [hunk 11]
```diff
@@ -45,11 +67,20 @@ func NewClientFromPublishSettingsFile(publishSettingsFilePath string) (Client, e
...
+	if subscriptionID == "" {
+		return client, errors.New("azure: subscription ID required")
+	} else if managementCertificatePath == "" {
+		return client, errors.New("azure: management certificate path required")
+	} else if baseURL == "" {
+		return client, errors.New("azure: base URL required")
+	}
+
+	client.baseURL = baseURL
+	err := client.importPublishSettings(subscriptionID, managementCertificatePath)
```
Comment (ahmetalpbalkan) [inline azure/client.go]: You can directly return this line

File: azure/virtualmachine/client.go [hunk 12]
```diff
@@ -73,10 +73,7 @@ func (self VirtualMachineClient) CreateAzureVM(azureVMConfiguration *Role, dnsNa
 		return err
 	}
 
-	err = self.client.WaitAsyncOperation(requestId)
-	if err != nil {
-		return err
-	}
+	self.client.WaitAsyncOperation(requestId)
```
Comment (ahmetalpbalkan) [inline azure/virtualmachine/client.go]: This must be already fixed. If not, please submit another pr

File: azure/client.go [hunk 13]
```diff
@@ -34,10 +37,29 @@ func NewAnonymousClient() Client {
...
+func NewClientFromPublishSettingsFile(publishSettingsFilePath string, baseURL string) (Client, error) {
 	client := Client{}
+	if publishSettingsFilePath == "" {
+		return client, errors.New("azure: publishsettings file path required")
+	} else if baseURL == "" {
+		return client, errors.New("azure: base URL required")
+	}
+
+	client.baseURL = baseURL
 	err := client.importPublishSettingsFile(publishSettingsFilePath)
```
Comment (ahmetalpbalkan) [inline azure/client.go]: You can directly return this.

File: azure/client.go [hunk 14]
```diff
@@ -2,10 +2,12 @@ package azure
 
 import (
 	"encoding/xml"
+	"errors"
 	"fmt"
 )
 
 const (
+	azureManagementDNSName                = "https://management.core.windows.net"
```
Comment (ahmetalpbalkan) [inline azure/client.go]: Better name: defaultAzureManagementURL

**Comment by anweiss** [comment]
fixed the rebase issue (hunk kept getting staged for some reason)...modified return statements and updated constant to more friendly name. slowly getting the hang of these conventions :)

File: azure/client.go [hunk 15]
```diff
@@ -34,24 +37,56 @@ func NewAnonymousClient() Client {
...
+		return client, errors.New("azure: base URL required")
+	}
+
 	err := client.importPublishSettingsFile(publishSettingsFilePath)
 	if err != nil {
 		return client, err
 	}
-	return client, nil
+
+	return Client{
```
Comment (ahmetalpbalkan) [inline azure/client.go]: this is particularly bad because if somebody sets `client.XXX = YYY`, you'll construct a new client :) Also have you noticed that these `return Client{baseURL:baseURL}` are duplicate? Maybe a new method? :wink:

Comment (anweiss) [inline azure/client.go]: lol, yep, missed that one...will update that

Comment (ahmetalpbalkan) [inline azure/client.go]: it's not entirely bad. I'm also fine with previous version like you set baseURL at the end. but it's probably better to have a `newClient(baseURL)` method and maybe even return `*Client` to get rid of being have to declare `var client Client`...

Comment (anweiss) [inline azure/client.go]: So returning a pointer was what I was thinking initially based on other API SDKs I've perused, but didn't want to do too much re-factoring...I guess at this point it wouldn't hurt

Comment (ahmetalpbalkan) [inline azure/client.go]: We actually recently decided to not to pass *Client to management libraries because then they can modify it. However if we return *Client here, then we expect people to initialize clients like virtualmachine.NewClient(*client) which is vague because they need to figure out deferencing it with `*`. It doesn't really sound idiomatic.

**Comment by jeffmendoza** [comment]
Can we leave the existing factories unchanged?

**Comment by anweiss** [comment]
We could...although, I feel like the client factories start to become unwieldy as we add more of them to support different parameter sets

**Comment by ahmetalpbalkan** [comment]
@anweiss +1. as we develop this the factory method will just keep getting bigger. it makes sense to at least have one very simple factory and one with multiple parameters or a Config object.

**Comment by jeffmendoza** [comment]
Sorry, I mean the function signatures, not the implementation.

**Comment by ahmetalpbalkan** [comment]
@jeffmendoza yep I'm aware. But we already broke everything. I thought it would be fine. Any specific reasons to not to break this?

Maybe we can keep NewClient same and add NewClientFromConfig.

**Comment by anweiss** [comment]
I'm not sure I like the way the publishsettings file/management certs are imported either...which impacts the design of these clients. Currently, the `importPublishSettingsFile` function simply takes the first subscription tag and assumes it to be the active. We may want to modify this function to accept both a publishsettings file path and subscription ID...that we can eliminate the redundancy of the `importPublishSettings` function and simplify the client model.

**Comment by ahmetalpbalkan** [comment]
@anweiss +1 and actually we shouldn't ask for `publishSettingsFilePath`, user should be responsible of loading that somewhere and just passing the cert object or []byte of the file to us. :smile: Many other client libraries we have also accept .publishSettings file in the constructur but mostly their common use case is `NewClient(subscriptionId, subscriptionCert, certPassword)`

**Comment by jeffmendoza** [comment]
Ah ok, makes sense. I think keeping the "first found" publishsettings file options would be good for some cases, but also specifying which sub to use would be a great addition.

Agree with taking in a byte array for these file.

**Comment by anweiss** [comment]
`SubscriptionKey` looks to be the private key as part of a pair of PEM encoded data as [passed](https://github.com/MSOpenTech/azure-sdk-for-go/blob/f3310618eb676675085a4839a3f8d5abad331da4/management/http.go#L108) to the `X509KeyPair` function. In this case, the public/private keys are the same.

**Comment by anweiss** [comment]
Going to close and re-open this PR to clean things up

**Comment by ahmetalpbalkan** [comment]
@anweiss I hate when people do that unless absolutely necessary :-)

**Comment by anweiss** [comment]
lol, I screwed up my rebase with the merged commits from the package refactor...so my commits in this PR kept including changes that were already merged in to master...hence the new branch
